### PR TITLE
feat: add user endpoints to new api

### PR DESCRIPTION
**Pull request by ybt195**
## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: proto/src/determined/user/v1/user.proto [hunk 1]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+
+package determined.user.v1;
+option go_package = "github.com/determined-ai/determined/proto/pkg/userv1";
+
+// User is an account in the determined cluster.
+message User {
+    // The id of the user.
+    int32 id = 1;
```
Comment (brain-good) [inline proto/src/determined/user/v1/user.proto:9]: numbers!!

File: master/internal/db/database.go [hunk 2]
```diff
@@ -1,20 +1,11 @@
...
-// that is already taken.
-type ErrDuplicateUser struct {
-	Username string
-}
-
-func (s ErrDuplicateUser) Error() string {
-	return fmt.Sprintf("user with username %q already exists", s.Username)
-}
+// ErrDuplicateRecord is returned when trying to create a row that already exists.
+var ErrDuplicateRecord = errors.New("duplicate")
```
Comment (brain-good) [inline master/internal/db/database.go:11]: blocking: make slightly clearer

File: master/internal/api_auth.go [hunk 3]
```diff
@@ -28,16 +28,21 @@ func (a *apiServer) Login(
...
 
 func (a *apiServer) CurrentUser(
 	ctx context.Context, _ *apiv1.CurrentUserRequest) (*apiv1.CurrentUserResponse, error) {
-	_, _, err := grpc.GetUser(ctx, a.m.db)
+	user, _, err := grpc.GetUser(ctx, a.m.db)
 	if err != nil {
 		return nil, err
 	}
-	return &apiv1.CurrentUserResponse{}, nil
+	fullUser, err := getUser(a.m.db, user.Username)
```
Comment (brain-good) [inline master/internal/api_auth.go:44]: non-blocking: i'm just going to assume this is needed

File: master/internal/api_user.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/grpc"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/userv1"
+)
+
+var errUserNotFound = status.Error(codes.NotFound, "user not found")
```
Comment (brain-good) [inline master/internal/api_user.go:17]: praise: nice

File: master/internal/api_user.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+func (a *apiServer) GetUser(
+	_ context.Context, req *apiv1.GetUserRequest) (*apiv1.GetUserResponse, error) {
+	fullUser, err := getUser(a.m.db, req.Username)
+	return &apiv1.GetUserResponse{User: fullUser}, err
+}
+
+func (a *apiServer) PostUser(
+	ctx context.Context, req *apiv1.PostUserRequest) (*apiv1.PostUserResponse, error) {
+	if err := grpc.ValidateRequest(
+		func() (bool, string) { return req.User != nil, "no user specified" },
```
Comment (brain-good) [inline master/internal/api_user.go:90]: non-blocking: are we sure we don't to pass an error object to the request? That way we can have better typed errors potentially in the future?

File: master/internal/api_user.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+		func() (bool, string) { return req.User.Username != "", "no username specified" },
+	); err != nil {
+		return nil, err
+	}
+	curUser, _, err := grpc.GetUser(ctx, a.m.db)
+	if err != nil {
+		return nil, err
+	}
+	if !curUser.Admin {
+		return nil, grpc.ErrPermissionDenied
```
Comment (brain-good) [inline master/internal/api_user.go:100]: question: I feel like we should move this to be the first thing we check?

File: master/internal/api_user.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+		}
+	}
+
+	switch err = a.m.db.AddUser(user, agentUserGroup); {
+	case err == db.ErrDuplicateRecord:
+		return nil, status.Error(codes.InvalidArgument, "user already exists")
+	case err != nil:
+		return nil, err
+	}
+	fullUser, err := getUser(a.m.db, req.User.Username)
```
Comment (brain-good) [inline master/internal/api_user.go:126]: question: seems odd to have to get the user back from the db after putting it in there, is this avoidable?

File: master/internal/api_user.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func (a *apiServer) SetUserPassword(
+	ctx context.Context, req *apiv1.SetUserPasswordRequest) (*apiv1.SetUserPasswordResponse, error) {
+	curUser, _, err := grpc.GetUser(ctx, a.m.db)
+	if err != nil {
+		return nil, err
+	}
+	if !curUser.Admin && curUser.Username != req.Username {
+		return nil, grpc.ErrPermissionDenied
```
Comment (brain-good) [inline master/internal/api_user.go:137]: same on checking permission first

Comment (ybt195) [reply master/internal/api_user.go:137]: Already the first check.

File: master/internal/api_user.go (hunk 7 above)
Comment (ybt195) [reply master/internal/api_user.go:126]: I am planning on cleaning up the database queries and interface after this project is done. For now, I am keeping it as is.

File: master/internal/api_user.go (hunk 8 above)
Comment (brain-good) [reply master/internal/api_user.go:137]: we're calling `GetUser` first, but maybe that's fine 🤷‍♂️

Comment (ybt195) [reply master/internal/api_user.go:129]: GetUser returns an error that needs to be handled first. The current user returned might not be a valid value if an error is thrown (e.g. there is no user logged in).

File: master/internal/api_user.go (hunk 5 above)
Comment (ybt195) [reply master/internal/api_user.go]: The grp error library takes strings as input, not errors

File: master/internal/db/database.go [hunk 9]
```diff
@@ -8,4 +8,4 @@ import (
 var ErrNotFound = errors.New("not found")
 
 // ErrDuplicateRecord is returned when trying to create a row that already exists.
-var ErrDuplicateRecord = errors.New("duplicate")
+var ErrDuplicateRecord = errors.New("duplicate row already exists")
```
Comment (brain-good) [inline master/internal/db/database.go:11]: Maybe just "row already exists"

**Pull request merged**